tools/update_glibc.zig

```diff
@@ -0,0 +1,333 @@
+    // Here we create a mapping of function name to list of versions.
+    for (abi_lists) |*abi_list, abi_index| {
+        const kv = target_functions.get(@ptrToInt(abi_list)).?;
+        const fn_vers_list = &kv.value.fn_vers_list;
+        for (kv.value.list.toSliceConst()) |*ver_fn| {
+            const gop = try fn_vers_list.getOrPut(ver_fn.name);
+            if (!gop.found_existing) {
+                gop.kv.value = std.ArrayList(usize).init(allocator);
+            }
+            const ver_index = global_ver_set.get(ver_fn.ver).?.value;
+            if (std.mem.indexOfScalar(usize, gop.kv.value.toSliceConst(), ver_index) == null) {
+                try gop.kv.value.append(ver_index);
+            }
+        }
+    }
+
+    {
+        const abilist_txt_path = try fs.path.join(allocator, [_][]const u8{ glibc_out_dir, "abi.txt" });
+        const abilist_txt_file = try fs.File.openWrite(abilist_txt_path);
+        defer abilist_txt_file.close();
+        var buffered = std.io.BufferedOutStream(fs.File.WriteError).init(&abilist_txt_file.outStream().stream);
+        const abilist_txt = &buffered.stream;
+
+        // first iterate over the abi lists
+        for (abi_lists) |*abi_list, abi_index| {
+            const fn_vers_list = &target_functions.get(@ptrToInt(abi_list)).?.value.fn_vers_list;
+            for (abi_list.targets) |target, it_i| {
+                if (it_i != 0) try abilist_txt.writeByte(' ');
+                try abilist_txt.print("{}-linux-{}", @tagName(target.arch), @tagName(target.abi));
+            }
+            try abilist_txt.writeByte('\n');
+            // next, each line implicitly corresponds to a function
+            for (global_fn_list) |name| {
+                const kv = fn_vers_list.get(name) orelse {
+                    try abilist_txt.writeByte('\n');
+                    continue;
+                };
+                for (kv.value.toSliceConst()) |ver_index, it_i| {
+                    if (it_i != 0) try abilist_txt.writeByte(' ');
+                    try abilist_txt.print("{d}", ver_index);
+                }
+                try abilist_txt.writeByte('\n');
+            }
+        }
+
+        try buffered.flush();
+    }
+}
+
+pub fn strCmpLessThan(a: []const u8, b: []const u8) bool {
+    return std.mem.compare(u8, a, b) == .LessThan;
+}
+
+pub fn versionLessThan(a: []const u8, b: []const u8) bool {
+    const sep_chars = "GLIBC_.";
+    var a_tokens = std.mem.tokenize(a, sep_chars);
+    var b_tokens = std.mem.tokenize(b, sep_chars);
+
+    while (true) {
+        const a_next = a_tokens.next();
+        const b_next = b_tokens.next();
+        if (a_next == null and b_next == null) {
+            return false; // equal means not less than
+        } else if (a_next == null) {
+            return true;
+        } else if (b_next == null) {
+            return false;
+        }
+        const a_int = fmt.parseInt(u64, a_next.?, 10) catch unreachable;
```

Comment:
Should this be `catch @panic()`? Is there anything to prevent someone compiling this in release-fast mode?

Reply:
Although I didn't intend for this code to ever be run in non-debug mode, it probably should do as you say, just to set a good example for people who are reading zig source and expecting it to be idiomatic. Same goes for the zig build system implementation.

Or, there is another option...

```zig
comptime{
    assert(builtin.mode == .Debug); // this tool is coded for debug mode only
}
```

Reply:
Yep either of those would be solutions.

If you lead by example you might think of new syntax to make the "right" thing easier to do and the "bad" thing harder to do.